### PR TITLE
Split QuantumInference into analytic and sampled subclasses

### LINKED ISSUE #225: Refactor QNN
Right now it is difficult to specify analytic expectations versus sampled expectations.

### PR DESCRIPTION
**Pull request by zaqqwerty**
First separation, part of #225

### CODE REVIEW

File: qhbmlib/inference/qnn.py [hunk 1]
```diff
@@ -26,136 +26,167 @@
...
+  """Analytic methods for inference on QuantumCircuit objects.
+
+  This class uses the TensorFlow Quantum `Expectation` layer to compute
+  expectation values of observables.  It uses adjoint gradients to compute
+  the derivatives of those expectation values.
+  """
+
+  def __init__(self,
+               input_circuit: circuit.QuantumCircuit,
+               name: Union[None, str] = None):
```
Comment (farice) [inline qhbmlib/inference/qnn.py:97]: could you summarize this piece of context in the docstrings which justifies why we now always use adjoint diff for analytic inference and parameter shift for sampled:

> Specifying alternative differentiators for the 'noiseless' backend, such as ParameterShift, only has the effect of increasing computational overhead, since the expectation values are exact irrespective of the chosen differentiator

Comment (zaqqwerty) [reply qhbmlib/inference/qnn.py:104]: Done, let me know if it clarifies things

**Review by farice — APPROVED** [review]
